### PR TITLE
Search V2.1

### PR DESCRIPTION
**Pull request by mikkpokk**

### CODE REVIEW

File: app/Http/Controllers/V2FrontpageController.php [hunk 1]
```diff
@@ -12,14 +13,16 @@ public function index()
     {
         $loggedUser = auth()->user();
 
-        $flights = Content::getLatestItems('flight', 9);
+        $flights = Content::getLatestItems('flight', 9, 'id');
```
Comment (kristjanjansen) [inline app/Http/Controllers/V2FrontpageController.php]: - Why sort on id? Performance?
- Lets split it into separate PR **Frontpage / destination speed improvements**

File: app/Http/Regions/DestinationBar.php [hunk 2]
```diff
@@ -2,26 +2,49 @@
...
+        //dd($parentDestinations);
 
         return component('DestinationBar')
             ->is($is)
             ->with('title', $destination->vars()->shortName)
             ->with('route', route('destination.show', [$destination]))
             ->with('parents', region(
                 'DestinationParents',
-                $destination->getAncestors()->onlyLast($parentLength > 25 ? 1 : 2),
+                $parentDestinations,
```
Comment (kristjanjansen) [inline app/Http/Regions/DestinationBar.php]: - Do we really need this magick when we could just keep the human-readable Baum API and wrap it into Cache::remember?
- Lets split it into separate PR **Frontpage / destination speed improvements**

File: app/Http/Regions/DestinationParents.php [hunk 3]
```diff
@@ -6,14 +6,18 @@ class DestinationParents
 {
     public function render($parents, $short = false)
     {
+        $loopCount = 0;
+        $parentsCount = $parents->count();
+
         return component('Meta')
-            ->with('items', $parents->map(function ($parent) use ($short) {
+            ->with('items', $parents->map(function ($parent) use ($short, &$loopCount, $parentsCount) {
```
Comment (kristjanjansen) [inline app/Http/Regions/DestinationParents.php]: - Add a comment what does $loopCount do and why we need it
- Lets split it into separate PR **Frontpage / destination speed improvements**

**Comment by mikkpokk** [comment]
Search is not ready yet, but you can test it. For testing you have to run these commands first:
 - gulp
 - php artisan search:index

**Comment by mikkpokk** [comment]
#1380

File: app/Console/Commands/ConvertSearchable.php [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+                    $key = 'u'.$index->user_id;
+                } else {
+                    $key = null;
+                }
+
+                if ($key && isset($index_fields[$key])) {
+                    if (! $index->title) {
+                        $index->title = 'null';
+                    } else {
+                        $index->title = DB::getPdo()->quote($index->title);
```
Comment (kristjanjansen) [inline app/Console/Commands/ConvertSearchable.php]: Why we need PDO here?

File: app/Console/Commands/ConvertSearchable.php [hunk 5]
```diff
@@ -0,0 +1,288 @@
+<?php
```
Comment (kristjanjansen) [inline app/Console/Commands/ConvertSearchable.php]: Tell me again why not we do it on demand, when Eloquent emits crud events? I got the point of having a command for initial seeding but is it neccessary later?

File: app/Console/Commands/ConvertSearchable.php [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        }
+
+        $index_count = count($index_fields);
+        if ($index_count > 0) {
+            $this->save($index_fields, $index_count);
+        } else {
+            $this->info(' + Nothing to save');
+        }
+
+        $this->info('Command finished after '.round(microtime(true) - $start, 4).' seconds');
```
Comment (kristjanjansen) [inline app/Console/Commands/ConvertSearchable.php]: Remove debug

File: app/Utils/utils.php [hunk 7]
```diff
@@ -5,6 +5,19 @@
 use Illuminate\Support\Facades\URL;
 use Illuminate\Support\Facades\Redirect;
 
+function full_text_safe($string)
```
Comment (kristjanjansen) [inline app/Utils/utils.php]: Have a short comment for the purpose of the function

File: resources/views/v2/components/FormSearchField/FormSearchField.css [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    align-items: center;
+    position: relative;
+    width: 100%;
+}
+
+    .FormSearchField .FormSearchField__input {
+        -webkit-transition: all 0.15s linear;
+        -moz-transition: all 0.15s linear;
+        -o-transition: all 0.15s linear;
+        transition: all 0.15s linear;
```
Comment (kristjanjansen) [inline resources/views/v2/components/FormSearchField/FormSearchField.css]: no need for vendor prefixes, we have autoprefix postcss

File: resources/views/v2/components/FormSearchField/FormSearchField.css [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    .FormSearchField__input:hover,
+    .FormSearchField__input:focus {
+        color: $black;
+    }
+
+        .FormSearchField:hover .FormSearchField__input,
+        .FormSearchField__input:focus {
+            background: rgb(255, 255, 255);
```
Comment (kristjanjansen) [inline resources/views/v2/components/FormSearchField/FormSearchField.css]: $white

File: resources/views/v2/components/FrontpageSearch/FrontpageSearch.css [hunk 10]
```diff
@@ -1,48 +1,125 @@
...
-    .FrontpageSearch__icon {
-        display: flex;
-        align-items: center;
-        cursor: pointer;
-        margin-right: $margin-xs;
+    .FrontpageSearch * {
+        -webkit-transition: all 0.15s linear;
+        -moz-transition: all 0.15s linear;
+        -o-transition: all 0.15s linear;
+        transition: all 0.15s linear;
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearch/FrontpageSearch.css]: No vendor prefixes

File: resources/views/v2/components/FrontpageSearch/FrontpageSearch.css [hunk 11]
```diff
@@ -1,48 +1,125 @@
...
+    padding: 5px 40px $padding-xxs 50px;
+    color: $black;
+    border: 0;
+    width: 100%;
+    height: 50px;
+}
+
+    .FrontpageSearch:hover .FrontpageSearch__input,
+    .FrontpageSearch__input:focus {
+        background: rgb(255, 255, 255);
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearch/FrontpageSearch.css]: $white

File: resources/views/v2/components/FormSearchField/FormSearchField.css [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+        .FormSearchField ::-webkit-input-placeholder { /* WebKit, Blink, Edge */
```
Comment (kristjanjansen) [inline resources/views/v2/components/FormSearchField/FormSearchField.css]: Can we collect these somehow or get rid of vendor prefixes totally? Verify the compiled css is autoprefix up the job

File: resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue [hunk 13]
```diff
@@ -1,73 +1,107 @@
...
+            },
             search: function() {
-                window.location = '/search?q=' + this.keywords
+                this.loading = true
+                if (this.searchTimeout) {
+                    clearTimeout(this.searchTimeout)
+                    this.searchTimeout = null
+                }
+
+                this.searchTimeout = setTimeout(() => { this.performSearch() }, 150);
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue]: try lodash.throttle

File: resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue [hunk 14]
```diff
@@ -1,73 +1,107 @@
...
+                                        this.lastRequest.abort()
+                                    }
+
+                                    this.lastRequest = xhr
+                                }
+                            }).then(
+                                (res) => {
+                                    this.results = res.data
+                                    this.loading = false
+                                }, function(error) {
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue]: error =>

File: resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue [hunk 15]
```diff
@@ -1,73 +1,107 @@
...
             },
-            toggleActive: function() {
-                this.active = ! this.active
+            performSearch: function() {
+                if (this.query && this.query != '' && String(this.query).length > 2 && ! this.enterPressed) {
+                    if (this.query != this.lastKeyword) {
+                        this.$http
+                            .get('/search/ajaxsearch/?q=' + this.query + '&_t=' + Date.now(), {
+                                before: function (xhr) {
+                                    if (this.lastRequest) {
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearch/FrontpageSearch.vue]: if using _.throttle this magick is not neccessary

File: resources/views/v2/components/FrontpageSearchItem/FrontpageSearchItem.css [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+.FrontpageSearchItem {
+    float: left;
+    width: 100%;
+    color: $blue;
+    padding-left: calc($padding-sm * 4);
+    margin-top: $padding-xs;
+    font-size: $font-size-sm;
+}
+
+    .FrontpageSearchItem__badge {
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearchItem/FrontpageSearchItem.css]: Do we really need the badges? Is it Haiku code?

File: resources/views/v2/components/FrontpageSearchItem/FrontpageSearchItem.vue [hunk 17]
```diff
@@ -0,0 +1,24 @@
+<template>
+
+    <a :href="result.route" class="FrontpageSearchItem" :class="isclasses">
+        {{ result.title }}
+         <span :class="'FrontpageSearchItem__badge FrontpageSearchItem__badge--' + (result.badge_color ? result.badge_color : 'orange')" v-if="result.badge">
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearchItem/FrontpageSearchItem.vue]: - Do we need the badges at all?
- Why not we use isclasses for dynamic coloring? I know its awkward with vue->vue but this is what we have not. Lets declare colors later as JS so they can be used both from CSS and JS

File: resources/views/v2/components/FrontpageSearchRow/FrontpageSearchRow.css [hunk 18]
```diff
@@ -0,0 +1,66 @@
+@import "variables";
```
Comment (kristjanjansen) [inline resources/views/v2/components/FrontpageSearchRow/FrontpageSearchRow.css]: lot of $non-variables and hardcoded px values. Can be followed up later as cleanup PR

File: resources/views/v2/components/HeaderTabs/HeaderTabs.css [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    position: relative;
+    top: $padding-lg;
+    top: $padding-sm ?if $mobile;
+}
+
+    ul.HeaderTabs * {
+        -webkit-transition: all 0.15s linear;
+        -moz-transition: all 0.15s linear;
+        -o-transition: all 0.15s linear;
+        transition: all 0.15s linear;
```
Comment (kristjanjansen) [inline resources/views/v2/components/HeaderTabs/HeaderTabs.css]: no prefixes

File: app/Utils/utils.php (hunk 7 above)
Comment (mikkpokk) [inline app/Utils/utils.php]: https://dev.mysql.com/doc/refman/5.7/en/fulltext-boolean.html

Otherwise:
a) we ignore words with "-", "_", "&", "%", "(", ")" etc..
b) MySQL eroor

File: app/Console/Commands/ConvertSearchable.php (hunk 6 above)
Comment (mikkpokk) [inline app/Console/Commands/ConvertSearchable.php]: Done
